### PR TITLE
Readme tweaks

### PR DESCRIPTION
**Pull request by vweevers**
Do not merge yet.

### CODE REVIEW

**Comment by vweevers** [comment]
I'm trying to make sense of `encoding.buffer = true` but woa..

**Comment by vweevers** [comment]
It used to mean "retrieve values from the underlying store as a Buffer, to avoid copying". If an encoding had `buffer: true` then leveldown would see `_get(key, { asBuffer: true })`. Right?

But if I follow current code paths I don't see this still being true.

**Comment by vweevers** [comment]
> It used to mean "retrieve values from the underlying store as a Buffer, to avoid copying". If an encoding had `buffer: true` then leveldown would see `_get(key, { asBuffer: true })`. Right?

Wrong. The other way around. If `asBuffer` is false, then leveldown returns a string, instead of a Buffer that would then be stringified anyway (by an encoding). Also it looks like a memcpy does happen for Buffers since https://github.com/Level/leveldown/commit/b7272db9f195a1070f4cdd15a99216ddc08f86eb, so that's a double cost when Buffers are not needed.

So @ralphtheninja @juliangruber could it be that we broke the `asBuffer` optimization somewhere along the way? Or am I just completely lost? This test fails:

<details><summary>click to expand</summary>

```js
'use strict'

const test = require('tape')
const levelup = require('levelup')
const leveldown = require('leveldown')
const encode = require('encoding-down')

test('json encoding retrieves a string from leveldown', function (t) {
  const down = leveldown('./db')
  const enc = encode(down, { valueEncoding: 'json' })
  const db = levelup(enc)

  enc.codec.encodings.json.decode = function (value) {
    t.is(Buffer.isBuffer(value), false) // This fails

    return JSON.parse(value)
  }

  db.put('key', { value: 2 }, function (err) {
    t.error(err)

    db.get('key', { value: 2 }, function (err, value) {
      t.error(err)
      t.same(value, { value: 2 })

      db.close(t.end.bind(t))
    })
  })
})
```
</summary>

**Comment by ralphtheninja** [comment]
Good detective work. I need to spend a few hours on this to wrap my brain around it. That diff is huge too :/

**Comment by ralphtheninja** [comment]
@vweevers Can you make an issue for this to summarize what you have found so we don't lose track of it?

**Comment by vweevers** [comment]
>  That diff is huge as well :(

Sorry, I meant to [link to this specific line](https://github.com/Level/leveldown/commit/b7272db9f195a1070f4cdd15a99216ddc08f86eb#diff-0a0a90a8810da7685a5be3a1657da592R134).

File: README.md [hunk 1]
```diff
@@ -9,17 +9,118 @@
 [![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)
 [![npm](https://img.shields.io/npm/dm/encoding-down.svg)](https://www.npmjs.com/package/encoding-down)
 
+## Introduction
+
+Stores like [`leveldown`] can only store strings and Buffers. For a richer set of data types you can wrap such a store with `encoding-down`. It allows you to specify an *encoding* to use for keys and values independently. This not only widens the range of input types, but also limits the range of output types. The encoding is applied to all read and write operations: it encodes writes and decodes reads.
+
+Many encodings are builtin courtesy of [`level-codec`]. The default encoding is `utf8` which ensures you'll always get back a string. You can also provide a custom encoding like `bytewise` - or your own!
+
+## Builtin encodings
```
Comment (vweevers) [inline README.md]: Hmm I should move this table to `level-codec`.

File: README.md [hunk 2]
```diff
@@ -28,14 +129,22 @@ db.put('name', 'encoding-down', function (err) {
...
 ### `const db = require('encoding-down')(db[, options])`
 
-* `db` `abstract-leveldown` compatible db such as `leveldown`, `memdown`, `level-js` etc
-* `options.keyEncoding` (string) defaults to `'utf8'`
-* `options.valueEncoding` (string) defaults to `'utf8'`
+* `db` must be an [`abstract-leveldown`] compliant store
+* `options.keyEncoding` (string or object) defaults to `'utf8'`
+* `options.valueEncoding` (string or object) defaults to `'utf8'`
+
+`options` are passed to [`level-codec`](https://github.com/level/codec).
```
Comment (vweevers) [inline README.md]: Use reference for this link.

File: README.md [hunk 3]
```diff
@@ -28,14 +129,22 @@ db.put('name', 'encoding-down', function (err) {
...
+
+## Custom encoding
 
-`options` are passed to [`level-codec`](https://github.com/level/codec), see [supported encodings](https://github.com/Level/codec#encodings) for more information.
+*todo*.
```
Comment (vweevers) [inline README.md]: Document the format in `level-codec` (most importantly: `encoding.buffer`), then link to it from here.

**Comment by ralphtheninja** [comment]
@vweevers I can finish off the docs as well if you like.

**Comment by ralphtheninja** [comment]
@vweevers Maybe the API section should say something about `.keyEncoding` and `.valueEncoding` (https://github.com/Level/encoding-down/tree/readme/tweaks#const-db--requireencoding-downdb-options) since if the encoding is a string, then the encoding will be picked from `level-codec` and if it's an object it's a _custom_ encoding, which then nicely segways into the next section.

**Comment by vweevers** [comment]
> I can finish off the docs as well if you like, starting with the comments you already made.

@ralphtheninja I'm 90% done, will push to both repos tomorrow.

> Maybe the API section should say something about .keyEncoding and .valueEncoding (https://github.com/Level/encoding-down/tree/readme/tweaks#const-db--requireencoding-downdb-options) since if the encoding is a string, then the encoding will be picked from level-codec and if it's an object it's a custom encoding, which then nicely segways into the next section.

Nice one, will add that.

> Maybe also have a silly example of a custom encoding :)

My thoughts exactly, so I wrote one with `lexicographic-integer` and `async/await` because why not.

**Comment by ralphtheninja** [comment]
> @ralphtheninja I'm 90% done, will push to both repos tomorrow.

@vweevers I'm guessing you are fixing the badges as well ;)

**Comment by vweevers** [comment]
Will squash after review.

File: README.md [hunk 4]
```diff
@@ -28,14 +117,50 @@ db.put('name', 'encoding-down', function (err) {
...
+const level = require('level')
+const lexint = require('lexicographic-integer')
+
+async function main () {
+  const db = level('./db8', {
+    keyEncoding: {
+      // Hey, someone should publish this!
+      type: 'lexicographic-integer',
+      encode: (n) => lexint.pack(n, 'hex'),
+      decode: lexint.unpack,
```
Comment (vweevers) [inline README.md]: but haha, this doesn't work until #19 lands. Because if `unpack` receives a Buffer it won't parse it as hex.

Comment (ralphtheninja) [reply README.md]: Should be ready for merge now with tests for `.keyAsBuffer` and `.valueAsBuffer`

**Comment by ralphtheninja** [comment]
@vweevers Rebase to latest master?

**Comment by vweevers** [comment]
@ralphtheninja take over if you want (just needs squashing), I'm not available rest of the day

**Pull request merged**